### ISSUE TITLE
invalid `InvalidReturnStatement` when using `int` template

### ISSUE THREAD

**Issue by azjezz** [vimeo/psalm#11472]
Ref: https://psalm.dev/r/9375464f54

In the example above, `bar` should not emit any error, as `$a` is known to be an int, and as seen in `foo`, `int + 1` results in `int`.

The problem seems to come from here: https://github.com/vimeo/psalm/blob/cf420941d061a57050b6c468ef2c778faf40aee2/src/Psalm/Internal/Analyzer/Statements/Expression/BinaryOp/ArithmeticOpAnalyzer.php#L199-L200

where psalm does not check generic parameters; when analyzing arithematic expressions, it should unwrap the generic parameter to its constraint.

The solution should look something like this ( maybe worth having such a function as part of `TUnion`? ):

```php
$flat_mapper = function($atomic) {
  if ($atomic instanceof TGenericParameter) {
    return $atomic->constraint->types;
  } else {
    return [$atomic];
  }
};

$left_types = flat_map($left->getAtomicTypes(), $flat_mapper);
$right_types = flat_map($right->getAtomicTypes(), $flat_mapper);
```

It might also be worth looking into other binary operators and checking if they suffer from the same problem.
